File: Source/Sorting/counting_sort.js
const MAX_VALUE = 100000;

function countingSort(arr) {
    let i, j;
    let count = new Array(MAX_VALUE + 1).fill(0);
    let output = [];

    // Count the occurrence of each number
    for(i = 0; i < arr.length; i++) {
        count[arr[i]]++;
    }

    // Reconstruct the sorted array
    for(i = 0; i <= MAX_VALUE; i++) {
        for(j = 0; j < count[i]; j++) {
            output.push(i);
        }
    }

    return output;
}

File: Source/Sorting/counting_sort.c
#include <stdlib.h>

#define MAX_VALUE 100000

void counting_sort(int arr[], int n) {
    int i, j;
    int idx = 0;
    int* count = calloc(MAX_VALUE + 1, sizeof(int));

    // Count the occurrence of each number
    for(i = 0; i < n; i++) {
        count[arr[i]]++;
    }

    // Reconstruct the sorted array
    for(i = 0; i <= MAX_VALUE; i++) {
        for(j = 0; j < count[i]; j++) {
            arr[idx++] = i;
        }
    }

    free(count);
}
